text/0000-value-with-callback-return.md

```diff
@@ -0,0 +1,336 @@
+
+An important thing for any contract writer, is that no third party contract should be able to violate the invariants / make state inconsistent. Currently it's very hard to write a contract like this and we want to make sure developers don't need to worry about this.
+
+For example if you are building a fungible token contract `fun_token` with locks, if someone calls:
+```
+fun_token.lock(account_id, amount).then(|| {
+  assert(some_reason);
+})
+```
+
+This will lead to lock forever, because `fun_token` doesn't receive callback of failure of it's own callback.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+The proposed idea is to be able to attach a callback to the caller's callback.
+
+### Example:
+
+User `alice` calls exchange `dex` to swap token `fun` with token `nai` 
+
+Here is how it works now:
+- `alice` calls `dex`
+- `dex` calls `fun` and `nai` to lock corresponding balances. Attaches a callback back to `dex` to call `on_locks`
+- `fun` and `nai` locks balances within their contracts
+- `on_locks` on `dex` is called.
+    - If both locks succeeded, `dex` calls `fun` and `nai` to transfer funds.
+    - If one of the locks failed, `dex` calls the a token contract with the successful lock to unlock funds.
+   
+The issue is if the callback on `dex` fails for some reason, the tokens might remain locked.
+
+The proposal is:
+- `alice` calls `dex`
+- `dex` calls `fun` and `nai` to lock corresponding balances. Attaches a callback back to `dex` to call `on_locks`
+- `fun` and `nai` locks balances within their contracts.
+And also creates a new promise to unlock themselves. But instead they return it with `value_with_callback_return` (debatable name).
+This will attach a new promise to the result of `dex`'s callback (for each token).
+- `on_locks` on `dex` is called. `dex` can assert both locks succeeded.
+`dex` calls `fun` and `nai` to transfer funds. Attaches a new callback back to `dex`.
+- a new callback on `dex` is called. It can be noop.
+
+Two things to notice:
+1. `dex` no longer need to unlock tokens in case of lock failures, instead it can assert them.
+2. `dex` has to attach a callback to token transfer.
+
+### No need to unlock
+
+`dex` doesn't need to explicitly unlock tokens, because tokens can now attach a callback to unlock themselves.
+This callback going to executed when `on_locks` on `dex` finishes.
+So if `on_locks` method fails early on asserts, the unlock callbacks are going to be called. But only for successful locks.
+
+### Need to depend on token transfers
+
+This is a little more complicated. The reason `dex` needs to wait and depend on the token transfers is to avoid
+tokens from being unlocked before transfers are completed.
+
+If `dex` doesn't depend on transfers, the `unlock` might be executed before transfers, and someone might try to front-run it, so one of the transfers might fail.
```

Comment:
Understandable, but easy to forget to do it.
Can we somehow force this with API or automatically attach an all sink on success?

Reply:
In theory transfer receipts are going to arrive first and data receipts (for unlocks) later. But it only the case if there are no proxy in between. So it's not an option to depend on it.

I don't see a simple solution for this. It might be done in `near-bindgen` if within an external trait we have a decorator that indicates a potential lock.
The issue is the callback doesn't have a clue about locks. But maybe we can have a decorator on transferFrom to say you have to await for it. Either way it's not an all-case solution.

Reply:
But it doesn't solve the issue, that one unlock arrives before the transfer, while other transfer succeeds